gopass*: build with Go with CL417615

flokli opened the pull request:
gopass fails to build with 1.18.4 due to
https://github.com/gopasspw/gopass/issues/2280.

Apply the revert from upstream CL411617, CL417615.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

zowoq commented on pkgs/top-level/all-packages.nix:
```diff
@@ -2601,15 +2601,47 @@ with pkgs;
+  };
 
-  gopass-summon-provider = callPackage ../tools/security/gopass/summon.nix { };
+  # go 1.18 requires a newer Apple SDK to be build. See commit message for more details.
```
doesn't make sense, commit message doesn't include this detail.
```suggestion
  # go 1.18 requires a newer Apple SDK to be build.
```

zowoq reviewed (commented):
please remove me as commit author.

zowoq commented:
I don't agree with merging this but I assume you will merge it anyway.

flokli commented:
> I don't agree with merging this but I assume you will merge it anyway.

No :-D The main idea here is to unbreak the `gopass` attribute. We can use 1.18.4, but applying the backport queued for 1.18.5. That's probably still less obstrusive than a big rebuild due to applying it for all of go 1.18.4.

I'll update the PR.

zowoq commented:
Can you do this:

- apply the backport here for `gopass`
- apply the backport on `staging` for `go`
- revert the `gopass` from this PR backport on `staging`

flokli committed:
gopass*: build with Go with CL417615

gopass fails to build with 1.18.4 due to
https://github.com/gopasspw/gopass/issues/2280.

Apply the revert from upstream CL411617, CL417615.

Mic92 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -2601,15 +2601,47 @@ with pkgs;
+  # go 1.18 requires a newer Apple SDK to be build. See commit message for more details.
+  __buildGo1183Module = darwin.apple_sdk_11_0.callPackage ../development/go-modules/generic {
+    go = buildPackages.go_1_18.overrideAttrs (_: rec {
+      version = "1.18.3";
```
Would it not be easier to just build with go 1.17 rather than building a complete go compiler for one program?

flokli replied:
I think gopass needs go 1.18, due to the use of generics.

With the strategy from https://github.com/NixOS/nixpkgs/pull/182784#issuecomment-1193659652, we'll end up shipping the fix in go 1.18, before 1.18.5 is released.

Mic92 replied:
Nevermind. Gopass does not build with go 1.17.

flokli commented:
This now properly includes the revert for CL411617, CL417615.

Once that's merged, and master merged into staging, happy to send a PR there reverting this commit and shipping the patch in the go 1.18 expression itself.

zowoq commented:
@ofborg build gopass git-credential-gopass gopass-hibp gopass-jsonapi gopass-summon-provider

zowoq reviewed (approved):
Thanks.

flokli commented:
> Can you do this:
> 
>     * apply the backport here for `gopass`
> 
>     * apply the backport on `staging` for `go`
> 
>     * revert the `gopass` backport from this PR on `staging`

PR applying this to `go_1_18` directly, then reverting this commit is at https://github.com/NixOS/nixpkgs/pull/182839.
